### ISSUE TITLE
Custom derive attribute stripping is insufficient

### ISSUE THREAD

**Issue by jonas-schievink** [rust-lang/rust-analyzer#8434]
In

```rust
#[derive(Derive)]
#[cfg_attr(never, derive(Clone))]
struct S {
    #[cfg(never)]
    invisible_field: u8,
    #[cfg(not(never))]
    visible_field: u8,
}
```

The `Derive` macro should be passed the token stream corresponding to

```rust
struct S {
    #[cfg(not(never))]
    visible_field: u8,
}
```

Currently, we pass

```rust
#[cfg_attr(never, derive(Clone))]
struct S {
    #[cfg(never)]
    invisible_field: u8,
    #[cfg(not(never))]
    visible_field: u8,
}
```

Since we have to evalutate `cfg`s to do this correctly, I propose moving the `remove_derive_attrs` code to `hir_expand`.

**Comment by jonas-schievink** [comment]
Also, we should only be removing `#[derive]` attributes that come before the currently expanded derive macro, at least when https://github.com/rust-lang/rust/pull/84023 lands (see its tests)

**Comment by jonas-schievink** [comment]
Another relevant upstream PR: https://github.com/rust-lang/rust/pull/84110

**Comment by Arjentix** [comment]
Hello! Our team is struggling because of this bug. Do you plan to fix this? It's open for 2 years already.

**Comment by silver-ymz** [comment]
Same problem. Is there any progress on this issue?

**Comment by jeff-hiner** [comment]
This got worse for me about 3 or 4 weeks ago, specifically with `#[cfg(target_os = ...)]` fields. Maybe a regression? It's unfortunately making it impossible to see real errors in VSCode.

**Comment by ClementTsang** [comment]
Yeah, starting to see this a lot more for code where there's one or two fields that have a conditional field, and it's a bit annoying.

This can *kinda* be worked around in some cases by just manually deriving, but then that leads to clippy complaining for things like manually deriving `Default`.

**Comment by smalis-msft** [comment]
Any updates here? This would be a big usability boost IMO.

**Comment by izderadicka** [comment]
Same (or very similar) problem here with serde derive,   it causes  no-such-field error , it's caused by field, which is behind inactive feature, line 164,165. If I enable the feature in rust-analyzer settings that it's OK.
I'm almost sure that it did work some time ago - few month, so wondering why this old issue demostrated  just now.

![image](https://github.com/rust-lang/rust-analyzer/assets/2185733/95daf876-c6ac-4147-b396-b9770d33da2d)

**Comment by ssrlive** [comment]
OMG, a three-year-age bug.

**Comment by wyatt-herkamp** [comment]
I have started implementing a fix for this. 
![image](https://github.com/rust-lang/rust-analyzer/assets/11785959/94257791-da3c-4b12-9d4e-36ca9b736a4d)

@rustbot claim
